crates/futures/src/polyfill.rs

```diff
@@ -0,0 +1,208 @@
+ *
+ * Load this file in all agents that will use Atomics.waitAsync.
+ *
+ * Agents that don't call Atomics.waitAsync need do nothing special.
+ *
+ * Any kind of agent can wake another agent that is sleeping in
+ * Atomics.waitAsync by just calling Atomics.wake for the location being slept
+ * on, as normal.
+ *
+ * The implementation is not completely faithful to the proposed semantics: in
+ * the case where an agent first asyncWaits and then waits on the same location:
+ * when it is woken, the two waits will be woken in order, while in the real
+ * semantics, the sync wait will be woken first.
+ *
+ * In this polyfill Atomics.waitAsync is not very fast.
+ */
+
+/* Implementation:
+ *
+ * For every wait we fork off a Worker to perform the wait.  Workers are reused
+ * when possible.  The worker communicates with its parent using postMessage.
+ */
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use js_sys::{
+    encode_uri_component, Array, Atomics, Error, Function, Int32Array, JsString, Promise, Reflect,
+    SharedArrayBuffer, WebAssembly,
+};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use web_sys::{MessageEvent, Worker};
+
+macro_rules! console_log {
+    ($($t:tt)*) => (log(&format_args!($($t)*).to_string()))
+}
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_namespace = console)]
+    fn log(s: &str);
+}
+
+const DEFAULT_TIMEOUT: f64 = 10.0;
+
+const HELPER_CODE: &'static str = "
+onmessage = function (ev) {
+    try {
+        switch (ev.data[0]) {
+            case 'wait': {
+                let [_, ia, index, value, timeout] = ev.data;
+                console.log('wait event inside a worker');
+                let result = Atomics.wait(ia, index, value, timeout);
+                postMessage(['ok', result]);
+                break;
+            }
+            default: {
+                throw new Error('Wrong message sent to wait helper: ' + ev.data.join(','));
+            }
+        }
+    } catch (e) {
+        console.log('Exception in wait helper');
+        postMessage(['error', 'Exception']);
+    }
+}
+";
+
+thread_local! {
+    static HELPERS: RefCell<Vec<Rc<Worker>>> = RefCell::new(vec![]);
+}
+
+fn alloc_helper() -> Rc<Worker> {
+    HELPERS.with(|helpers| {
+        if let Some(helper) = helpers.borrow_mut().pop() {
+            return helper;
+        }
+
+        let mut initialization_string = "data:application/javascript,".to_owned();
+        let encoded: String = encode_uri_component(HELPER_CODE).into();
+        initialization_string.push_str(&encoded);
+
+        return Rc::new(Worker::new(&initialization_string).expect("Should create a Worker"));
+    })
+}
+
+fn free_helper(helper: &Rc<Worker>) {
+    HELPERS.with(move |helpers| {
+        helpers.borrow_mut().push(helper.clone());
+    });
+}
+
+pub fn wait_async(index: u32, value: i32) -> Result<Promise, JsValue> {
+    wait_async_with_timeout(index, value, DEFAULT_TIMEOUT)
```

Comment:
I think that instead of a `DEFAULT_TIMEOUT` you can pass `-1` here which indicates "wait forever"

Or at least I think that's the intented API of  `Atomics.waitAsync`

Reply:
In the docs it says that it defaults to `Infinity`, so I just made it as default. I didn't find any info about `-1` in the docs, and desided not to use it